tools/astarte_e2e/lib/astarte_e2e.ex

```diff
@@ -0,0 +1,168 @@
+# limitations under the License.
+#
+
+defmodule AstarteE2E do
+  @moduledoc """
+  Documentation for `AstarteE2E`.
+  """
+
+  use Application
+  require Logger
+
+  alias Astarte.Device
+  alias AstarteE2E.{Client, Config, Scheduler, Utils}
+
+  @standard_interface_path "priv/interfaces"
+
+  @type device_option ::
+          {:pairing_url, String.t()}
+          | {:realm, String.t()}
+          | {:device_id, String.t()}
+          | {:credentials_secret, String.t()}
+          | {:credential_storage, {module(), term()}}
+          | {:interface_provider, {module(), interfaces: list()}}
+          | {:handler, {module(), term()}}
+          | {:ignore_ssl_errors, boolean()}
+
+  @type client_option ::
+          {:url, String.t()}
+          | {:realm, String.t()}
+          | {:jwt, String.t()}
+          | {:device_id, String.t()}
+          | {:ignore_ssl_errors, boolean()}
+
+  @type device_options :: [device_option()]
+  @type client_options :: [client_option()]
+
+  @type interface_provider :: {module(), keyword()}
+
+  def start(_type, _args) do
+    with :ok <- Config.validate() do
+      opts = Config.astarte_e2e_opts!()
+
+      Supervisor.start_link(__MODULE__, opts, name: __MODULE__)
+    else
+      {:error, reason} ->
+        Logger.warn("Configuration incomplete. Unable to start process with reason: #{reason}.")
+        {:shutdown, reason}
+    end
+  end
+
+  def init(opts) do
+    children = [
+      AstarteE2EWeb.Telemetry,
+      {Device, device_opts(opts)},
+      {Client, client_opts(opts)},
+      {Scheduler, scheduler_opts(opts)}
+    ]
+
+    Supervisor.init(children, strategy: :one_for_one)
+  end
+
+  def work do
+    with {:ok, device_pid} <- fetch_device_pid(Config.realm!(), Config.device_id!()),
+         {:ok, interface_names} <- fetch_interface_names(),
+         :ok <- Device.wait_for_connection(device_pid) do
+      Enum.each(interface_names, fn interface_name ->
+        timestamp = :erlang.monotonic_time(:millisecond)
+
+        case interface_name do
+          "org.astarte-platform.e2etest.SimpleDatastream" ->
+            value = Utils.random_string()
+            path = "/correlationId"
+
+            Device.send_datastream(device_pid, interface_name, path, value)
+            :telemetry.execute([:astarte_end_to_end, :messages, :sent], %{}, %{})
+
+            :timer.sleep(1000)
+            Client.verify_device_payload(interface_name, path, value, timestamp)
+
+          "org.astarte-platform.e2etest.SimpleProperties" ->
+            value = Utils.random_string()
+            path = "/correlationId"
+
+            Device.set_property(device_pid, interface_name, path, value)
+            :telemetry.execute([:astarte_end_to_end, :messages, :sent], %{}, %{})
+
+            :timer.sleep(1000)
+            Client.verify_device_payload(interface_name, path, value, timestamp)
+        end
+      end)
+    end
+  end
+
+  defp device_opts(opts) do
```

Comment:
Regarding all the `opts` functions: you have a `Config` module that has all the options, then you mash them all together in `astarte_e2e_opts` and then you have to extract all the different option sets here.
I think it's better moving `device_opts`, `scheduler_opts` and `client_opts` (and also `standard_interface_provider` and the relative attribute, since they belong there) to the Config module so you can just call `Config.device_opts` and have a cleaner Application module